themes/qgrid/source/js/menu.js

```diff
@@ -15,41 +15,234 @@ function toggleVisibility(e) {
 	}
 }
 
-function filter(e) {
+function searchOnLoad() {
+	const active = document.querySelector('.active');
+	const search = getSearch();
+	if (active && search) {
+		pageTextSearch(search);
+		tagsSearch(search);
+		menuItemsSearch(search);
+	}
+}
+
+function searchOnHashChange() {
+	const search = getSearch();
+	updateMenuLinks(search);
+	pageTextSearch(search);
+	tagsSearch(search);
+	menuItemsSearch(search);
+}
+
+function onSearchChange(e) {
 	e.stopImmediatePropagation();
 	const input = e.target || e.srcElement;
-	const filter = input.value.toLowerCase();
+	const search = input.value.toLowerCase();
+	setSearch(search);
+}
+
+function setSearch(search) {
+	if (search) {
+		location.hash = 'search=' + search;
+	}
+	else {
+		location.hash = '';
+	}
+}
+
+function getSearch() {
+	const search = new RegExp(/^#search=/);
+	if (search.test(location.hash)) {
+		return location.hash.substr(search.source.length - 1).replace(/%20/g, ' ').toLowerCase();
+	}
+	return '';
+}
+
+function updateMenuLinks(search) {
 	const nav = document.getElementById('nav');
-	const li = nav.getElementsByTagName('li');
-
-	for (let i = 0; i < li.length; i++) {
-		const a = li[i].getElementsByTagName('a')[0];
-		if (a) {
-			if (a.text.toLowerCase().indexOf(filter) > -1) {
-				li[i].style.display = '';
-			} else {
-				li[i].style.display = 'none';
+	const menuItems = nav.getElementsByTagName('li');
+	for (let searchTarget of menuItems) {
+		const title = searchTarget.querySelector('.title');
+		if (title) {
+			let href = title.getAttribute('href');
+			const index = href.lastIndexOf('#search=');
+			if (index == -1 && search != '') {
+				href += '#search=';
 			}
+			else {
+				if (index > -1) {
+					href = href.substr(0, href.indexOf('=') + 1);
+				}
+				if (search == '') {
+					href = href.replace(/#search=/, '');
+				}
+			}
+			title.setAttribute('href', href + search);
+		}
+	}
+}
+
+function pageTextSearch(search) {
+	const h2 = document.getElementsByTagName('h2');
+	const p = document.getElementsByTagName('p');
+	for (let searchTarget of [...h2, ...p]) {
+		const a = searchTarget.getElementsByTagName('a')[0];
+		if (a && a.textContent) {
+			searchTarget = a;
+		}
+		highlightText(searchTarget, search);
+	}
+	for (let searchTarget of [...h2, ...p]) {
+		const index = searchTarget.textContent.toLowerCase().indexOf(search);
+		if (index > -1 && search != '') {
+			searchTarget.scrollIntoView();
+			return;
+		}
+	}
+}
+
+function tagsSearch(search) {
+	const active = document.querySelector('.active');
+	const nav = document.getElementById('nav');
+	const menuItems = nav.getElementsByTagName('li');
+	for (let searchMenuItem of menuItems) {
+		const tags = searchMenuItem.dataset.tag;
+		if (tags) {
+			const tag = tags.split(',');
+			for (let searchTag of tag) {
```

Comment:
we have tags, tag and searchTag in one place this is confusing

Reply:
for (let tag of tags.split(','))

dont need to have explicit variable in this case